tensorflow_federated/python/core/impl/executors/federating_executor.py

```diff
@@ -453,6 +451,221 @@ def validate_executor_placements(cls, executor_placements):
+    # cost from liner (with respect to the number of clients) to sub-linear.
+
+    # TODO(b/134543154): Expand this implementation to take advantage of the
+    # parallelism afforded by `merge`.
+    fed_ex = self.federating_executor
+
+    val = arg.internal_representation[0]
+    zero = arg.internal_representation[1]
+    accumulate = arg.internal_representation[2]
+    pre_report = await fed_ex._compute_intrinsic_federated_reduce(
+        FederatingExecutorValue(
+            anonymous_tuple.AnonymousTuple([(None, val), (None, zero),
+                                            (None, accumulate)]),
+            computation_types.NamedTupleType(
+                (val_type, zero_type, accumulate_type))))
+
+    py_typecheck.check_type(pre_report.type_signature,
+                            computation_types.FederatedType)
+    type_utils.check_equivalent_types(pre_report.type_signature.member,
+                                      report_type.parameter)
+
+    report = arg.internal_representation[4]
+    return await fed_ex._compute_intrinsic_federated_apply(
+        FederatingExecutorValue(
+            anonymous_tuple.AnonymousTuple([
+                (None, report), (None, pre_report.internal_representation)
+            ]),
+            computation_types.NamedTupleType(
+                (report_type, pre_report.type_signature))))
+
+  async def federated_sum(self, arg):
+    py_typecheck.check_type(arg.type_signature, computation_types.FederatedType)
+    zero, plus = tuple(await asyncio.gather(*[
+        executor_utils.embed_tf_scalar_constant(
+            self.federating_executor,
+            arg.type_signature.member,
+            0),
+        executor_utils.embed_tf_binary_operator(
+            self.federating_executor,
+            arg.type_signature.member,
+            tf.add)]))
+    return await self.federating_executor._compute_intrinsic_federated_reduce(
+        FederatingExecutorValue(
+            anonymous_tuple.AnonymousTuple([
+                (None, arg.internal_representation),
+                (None, zero.internal_representation),
+                (None, plus.internal_representation)
+            ]),
+            computation_types.NamedTupleType(
+                (arg.type_signature, zero.type_signature, plus.type_signature)))
+    )
+
+  async def federated_mean(self, arg):
+    # TODO check this logic, just switched SERVER to AGGREGATOR
+    arg_sum = await self.federating_executor._compute_intrinsic_federated_sum(arg)
+    member_type = arg_sum.type_signature.member
+    count = float(len(arg.internal_representation))
+    if count < 1.0:
+      raise RuntimeError('Cannot compute a federated mean over an empty group.')
+    child = self._get_child_executors(placement_literals.AGGREGATOR, index=0)
+    factor, multiply = tuple(await asyncio.gather(*[
+        executor_utils.embed_tf_scalar_constant(child, member_type,
+                                                float(1.0 / count)),
```

Comment:
don't know what the best solution is here, but it's kind of a weird mix between parts of the computation happening on the driver (computing the scaling factor) and parts on the aggregator (multiplication). from a security perspective you then ending up trusting both the aggregator and the driver (even if you currently have to anyway to some extent)

Reply:
Hadn't though of the scalar thing -- I took it very literally when we said everything except the aggregation happens elsewhere. I agree this is not ideal

Reply:
are you concerned with the division being done on the driver, or that the driver is the source of truth for the `count` variable? a workaround to fix the former seems like a lot of effort and unnecessary computation just to move that division over to the aggregator. the latter does not seem fixable -- even if we rearrange things so that `float(len(arg.internal_representation))` is computed on the aggregator placement, the driver is still the source of truth for the `arg`.

I'm not sure this is worth fixing, but maybe there's a better way than what I described. what do you think?

Reply:
I don't really know what would be best here. We're at the mercy of the driver no matter what. I think it's one of those things we'd definitely have to fix for FL/EL in deployment, but it does seem to require a change in architecture, and hence leaning towards not worth fixing here. In any case, happy to merge up this guy and let it simmer in the back for a while.

Technically speaking the computation on the aggregator could verify that it got exactly one input from a predefined clients -- yet with the current architecture the driver would supply this list, so we've just moved the problem.